### ISSUE TITLE
[bug] Abnormal loss fluctuation when resuming training from checkpoint

### ISSUE THREAD

**Issue by gaojingwei** [NVIDIA-NeMo/Megatron-Bridge#4565]
### Problem

I am performing SFT on a 236B model with DeepSeek V3 architecture. The training was interrupted at around step 25xx. I resumed training from step 2448, which is approximately at the 1.5th epoch. After resuming, the loss dropped to around 0.16 and remained at that level until 200 step before the end of the second epoch，then it raise to 0.2。 Before the interruption, the loss during the second epoch was consistently around 0.2.

I suspected that the data shuffling might be problematic when resuming from checkpoint. To verify this, I tried resuming from step 1632, which is exactly at the end of the first epoch — essentially restarting the second epoch from scratch. After resuming, the loss was around 0.23, which is significantly higher than the normal value of 0.2,  after 100 steps it fall to 0.2。

As shown in the figure below, the red-highlighted region indicates where we resumed training from the checkpoint at step 2448. The loss dropped from slightly below 0.2 to around 0.16. It then returned to approximately 0.2 near the end of the second epoch. At the start of the third epoch, the loss decreased to 0.1 and returned to normal behavior.

<img width="625" height="489" alt="Image" src="https://github.com/user-attachments/assets/d636f956-a5b5-484d-accb-0cc275b4e808" />

### Minimal repro

```shell
this is my script:

def create_finetune_config():
    model = AutoBridge.from_hf_pretrained("hf_model_path").to_megatron_provider(
            load_weights=False
        )
    return ConfigContainer(
        model= model,
        ddp=DistributedDataParallelConfig(
            check_for_nan_in_grad=True,
            grad_reduce_in_fp32=True,
            overlap_grad_reduce=True,
            overlap_param_gather=True,
            bucket_size=100000000,
        ),
        train=TrainingConfig(
            micro_batch_size=1,
            global_batch_size=GLOBAL_BATCH_SIZE,
            train_iters=MAX_STEP,

        ),
        optimizer=OptimizerConfig(
            optimizer="dist_muon",
            lr=2.4e-5,
            min_lr=2.4e-6,
            muon_extra_scale_factor = 0.2,
            weight_decay = 0.1,
            adam_beta1 = 0.9,
            adam_beta2 = 0.95,
            clip_grad = 1.0,
            bf16 = True,
            fp8_recipe = None,
            adam_eps = 1e-8,
            use_distributed_optimizer = False,      # True, ...muon must be false
            overlap_param_gather = True,            # keep same with ddp
        ),
        scheduler=SchedulerConfig(
            start_weight_decay=0.1,
            end_weight_decay=0.1,
            weight_decay_incr_style="constant",
            lr_decay_style="cosine",
            lr_warmup_iters=50,
        ),
        # Finetuning dataset instead of pretraining dataset
        dataset=FinetuningDatasetConfig(
            dataset_root="path_data_root",
            seq_length=SEQ_LENGTH,
            do_validation=False,
            do_test=False,
            # Optional: packed sequence support
            packed_sequence_specs=PackedSequenceSpecs(
                 packed_sequence_size=SEQ_LENGTH,
                packed_train_data_path=PACKED_TRAIN_DATA,
                packed_val_data_path=PACKED_VAL_DATA,
            ),
            dataset_kwargs = {"pad_to_max_length": True}
        ),
        tokenizer=TokenizerConfig(
            tokenizer_type="HuggingFaceTokenizer",
            tokenizer_model="tokenizer path",
        ),
checkpoint=CheckpointConfig(
            pretrained_checkpoint=RESTORE_PATH,
            save=CKPT_DIR,
            load=CKPT_DIR,
            save_interval=EVERY_N_TRAIN_STEPS,
            most_recent_k=100,
            async_save=True,
            fully_parallel_save=True,
            fully_parallel_load=True,
            dist_ckpt_strictness="log_unexpected",
        ),
        mixed_precision="bf16_mixed",
        rng=RNGConfig(seed=1234),
        comm_overlap=CommOverlapConfig(
            tp_comm_overlap=False,
        ),
        logger=LoggerConfig(
            log_interval=1,
            tensorboard_dir=CKPT_DIR,
        ),
    )
if __name__ == "__main__":
    cfg = create_finetune_config()
   
    cfg.model.fp8_recipe = None
    cfg.model.params_dtype = torch.float32
    cfg.optimizer.params_dtype = torch.float32
    cfg.optimizer.fp8_recipe = None

    # attention
    cfg.model.num_attention_heads = 128
    cfg.model.kv_channels = 40
    cfg.model.num_query_groups = 128
    cfg.model.num_layers = 60
    cfg.model.hidden_size = 5120
    cfg.model.ffn_hidden_size = 12288
    cfg.model.vocab_size = 152064

    #mtp
    cfg.model.mtp_num_layers = None
    cfg.model.mtp_loss_scaling_factor  = 0.1

    #moe
    cfg.model.num_moe_experts = 160
    cfg.model.moe_ffn_hidden_size = 1536
    cfg.model.moe_shared_expert_intermediate_size = 3072
    cfg.model.moe_layer_freq= [0] * 1 + [1] * 59
    cfg.model.moe_token_dispatcher_type = "flex"
    cfg.model.moe_enable_deepep = True
    cfg.model.moe_shared_expert_overlap = False           # moe_shared_expert_overlap only works with alltoall token dispatcher
    cfg.model.moe_aux_loss_coeff = 1e-4
    cfg.model.moe_z_loss_coeff = None
    cfg.model.moe_router_topk = 6
    cfg.model.moe_router_num_groups = 1
    cfg.model.moe_router_group_topk = 1
    cfg.model.moe_router_topk_scaling_factor = 2.827
    cfg.model.make_vocab_size_divisible_by = 512
    cfg.model.moe_router_score_function = "sigmoid"
    cfg.model.moe_router_enable_expert_bias = True
    cfg.model.moe_router_bias_update_rate = 1e-3
    cfg.model.moe_router_pre_softmax = True
    cfg.model.moe_grouped_gemm = True
    cfg.model.moe_router_load_balancing_type = "seq_aux_loss"
    cfg.model.moe_router_dtype = "fp32"

    #MLA
    cfg.model.multi_latent_attention=True
    cfg.model.kv_lora_rank = 512
    cfg.model.q_lora_rank = 1536
    cfg.model.qk_head_dim = 128
    cfg.model.qk_pos_emb_head_dim = 64
    cfg.model.v_head_dim = 128
    cfg.model.mscale = 1.0
    cfg.model.mscale_all_dim = 1.0

    # rope
    cfg.model.max_position_embeddings = SEQ_LENGTH
    cfg.model.seq_length = SEQ_LENGTH
    cfg.model.rotary_base = 5000000
    cfg.model.rope_type="rope"

    # recompute
    cfg.model.recompute_granularity = 'full'
    cfg.model.recompute_method = 'uniform'
    cfg.model.recompute_num_layers = 1

    # 3d parallel
    cfg.model.sequence_parallel=True
    #cfg.model.num_layers_in_first_pipeline_stage=8
    cfg.model.num_layers_in_last_pipeline_stage=4
    cfg.model.tensor_model_parallel_size=16
    cfg.model.pipeline_model_parallel_size=8
    cfg.model.expert_model_parallel_size=8
    cfg.model.expert_tensor_parallel_size=1
    cfg.model.virtual_pipeline_model_parallel_size=4
    cfg.model.context_parallel_size=2
    cfg.model.batch_p2p_comm=True

    # fusion
    cfg.model.attention_softmax_in_fp32 = True
    cfg.model.moe_permute_fusion = False
    cfg.model.apply_rope_fusion = False
    cfg.model.bias_activation_fusion = False
    cfg.model.bias_dropout_fusion = False
    cfg.model.masked_softmax_fusion = False
    cfg.model.gradient_accumulation_fusion = False
    cfg.model.cross_entropy_loss_fusion = False
    cfg.model.persist_layer_norm = False
    #cfg.model.params_dtype = torch.float32
    cfg.model.init_method_std = 0.006
    cfg.model.embedding_init_method_std = 0.006


    # when cp > 1，calculate_per_token_loss must be True
    cfg.model.calculate_per_token_loss = True if cfg.model.context_parallel_size > 1 else False
    cfg.model.sequence_parallel = True if cfg.model.tensor_model_parallel_size > 1 else False
    cfg.dataset.packed_sequence_specs.pad_seq_to_mult = cfg.model.context_parallel_size * 2 if  cfg.model.context_parallel_size > 1 else 1

    print_rank_0("-----all transformer config-------")
    print_rank_0(cfg.model)

    finetune(cfg, forward_step_func=forward_step)
```

### Expected behavior

The loss is expected to remain stable when resuming from a checkpoin

### Affected area

area:training

### Regression?

Yes

### Logs

```shell

```

**Comment by gaojingwei** [comment]
I debug code and find that：
in sft.py at GPTSFTPackedDataset：：_build_samples_mapping()  self.samples_mapping  is None
```
def _build_samples_mapping(self):
        if self.max_num_samples is not None:
            # custom samples mapping logic, following the format for unpacked sft dataset
            # Note: this is epoch-level shuffling, i.e. sampling without replacement until end of epoch, then repeat.
            # Unpacked dataset shuffles by sampling with replacement indefinitely.
            dataset_len = len(self.indexed_dataset)
            max_num_epochs = np.ceil(self.max_num_samples / dataset_len)
            indices = np.arange(dataset_len)[None, :].repeat(max_num_epochs, axis=0)
            [np.random.shuffle(x) for x in indices]
            self.samples_mapping = indices.reshape(1, -1).squeeze()[: self.max_num_samples]
            
            print(f"3333 global rank: {os.getenv("RANK", "0")} local_rank: {os.getenv("LOCAL_RANK")} dataset_len: {dataset_len}, {self.samples_mapping}")
        else:
            self.samples_mapping = None
```

at sampler.py  MegatronPretrainingBatchSampler::__iter__()， consumed_samples is not updated at every epoch and no shuffle at this function， maybe this is the root cause！
```
 batch = []
        # Last batch will be dropped if drop_last is True
        for idx in range(self.consumed_samples % self.total_samples, self.total_samples):
            batch.append(idx)
            if len(batch) == self._global_batch_size:
                # Distribute indices in interleaved fashion across ranks
                all_indices = [
                    batch[i]
                    for i in range(
                        self.data_parallel_rank,
                        self._global_batch_size,
                        self.data_parallel_size,
                    )
                ]
                assert len(all_indices) == self._global_batch_size_on_this_data_parallel_rank

                # Yield ALL indices at once (not split into microbatches)
                # The training loop will handle splitting after collation
                yield all_indices

                batch = []

        # Check the last partial batch and see if drop_last is set
        if len(batch) > 0 and not self.drop_last:
            # Distribute partial batch in interleaved fashion
            all_indices = [batch[i] for i in range(self.data_parallel_rank, len(batch), self.data_parallel_size)]
            if self.pad_samples_to_global_batch_size:
                num_pad = self._global_batch_size // self.data_parallel_size - len(all_indices)
                all_indices = all_indices + [-1] * num_pad

            # Yield ALL indices at once
            yield all_indices
```

**Comment by Achyuthan-S** [comment]
Thanks for the detailed write-up and especially for the debugging — your two pointers were exactly right.

**Root cause:** `MegatronPretrainingBatchSampler.__iter__` started from `consumed_samples % total_samples` but never advanced `consumed_samples`, unlike `MegatronPretrainingRandomSampler`. For fine-tuning the dataloader is wrapped in `cyclic_iter`, which re-calls `__iter__` on the same sampler instance every epoch — so the resume offset was re-applied on *every* pass. After a mid-epoch resume this permanently skipped the head of the dataset `[0, offset)` and replayed only the tail `[offset, total)`, re-training on the same samples and depressing the loss. Resuming exactly on an epoch boundary gives `offset = 0`, which is why that case looked normal — matching both of your observations (resume @2448 abnormal, resume @1632 fine).

Secondary factor: with `max_train_samples` unset (the default), the packed SFT dataset builds `samples_mapping = None`, so there was also no per-epoch reshuffle — every epoch served the same sequential order.

**Fix (#4601):** `MegatronPretrainingBatchSampler` is now epoch-aware — it advances `consumed_samples` as it yields so each re-iteration starts a fresh full epoch instead of replaying the tail, and adds deterministic, seed-/epoch-derived per-epoch reshuffling for the fine-tuning (`batch`) train dataloader (reshuffles every epoch and reproduces the same order on resume). Added regression tests for resume parity and per-epoch reshuffle.

Note: this changes the fine-tuning data order (deterministic and resume-reproducible), so SFT loss curves will differ slightly from the previous sequential ordering. Could you give the PR branch a try on your 236B run and confirm the resumed loss now stays continuous?
